### DISCUSSION TITLE
리팩토링 관련 논의

### DISCUSSION THREAD

**Discussion by KwonYeKyeong** [prgrms-web-devcourse/slam-server#125] [General]
이곳에서 리팩토링과 관련해 나누고 싶은 의견 있으면 적어주세요~

**Comment by KwonYeKyeong** [comment]
## api 수정하면서 고민하게 된 이슈 : 정적 팩토리 메서드 사용
### 정적 팩토리 메서드의 사용하게 된 이유(장점)
- 이름을 가질 수 있다
- 객체 생성을 캡슐화할 수 있다 -> entity에서 dto 변환시 내부 구현을 모르더라도 쉽게 변환할 수 있다, 예시) FollowingResponse.java의 `toResponse` 메서드
=> 좀 더 객체지향적으로 코드를 짤 수 있다고 생각했다. 그래서 생성자를 private으로 선언하게 되었다.
### 정적 팩토리 메서드를 사용함으로써 겪게된 문제점
- 테스트 코드 작성 시 객체 생성에 불편함을 겪게 되었다 -> 따라서 빌더패턴(`@Builder`)을 적용시켰다
=> 객체 생성에 있어 두가지 방식이 사용되고 있다. 좋은 코드라 할 수 있을까?
### 현재로써 내린 결론
entity 부분의 클래스에서는 빌더패턴을 사용하고, dto 부분의 클래스에서는 정적메서드 패턴을 사용하는 거 어떨까? (리팩토링 필요)

**Reply by locodingve** [reply]
넵~ 좋아요~ 

저는 이번에 notification 도메인 수정하면서, 이번 주에 수정하겠습니다. 
그리고서버 이전 완료된 후  chat 도메인은 수정하겠습니다.

**Reply by locodingve** [reply]
의견 감사합니다~ 멋져!!  👍

**Reply by KwonYeKyeong** [reply]
앗 공통적으로 적용하자는 뜻은 아니었어요~ 저는 저렇게 수정할꺼같다는 뜻이었습니다>_<

**Reply by locodingve** [reply]
하아! 그렇구나. 
그런데 공통으로 적용하는 게 좋을 것 같아요~ 
도메인 별 또는 개인 일정으로 적용하는 것이 어렵다면, 버전 2에는 적용이 되었으면 좋을 것 같아요~

각 도메인 별로 업무가 할당 되어, 각자의 스타일을 존중하지만 코드에 일관성은 있어야 할 것 같아요~

저도 강제가 아니라, 저의 생각을 말씀드린 것 입니다.

**Reply by locodingve** [reply]
젤리 ~ 
제가 젤리의 의견을 잘 못 알아 들었는지 궁금합니다. 
젤리가 말한 내용은 **한 클래스**에서 빌더 패턴과 정적팩토리 메소드를 함께 사용한 부분을 지적하신 건가요?

**Reply by locodingve** [reply]
위 의 질문과 별개로 아래의 의견을 드립니다. 

entity에는 정적팩토리 메서드를 쓰고, dto에는 빌더패턴을 쓰자고 하셨는데요. 
제가 작업하다가 보니까. dto 쪽에는 빌더 패턴으로 하기에는 클라이언트가 코드를 잘못 쓰면 큰일 아는 부분이있어서요. 

각자 판단 하에 정적팩토리 메서드 또는 빌더 패턴을 써도 될까요? (한 클래스에서 함께 사용한다는 의미가 아닙니다.)

**Reply by KwonYeKyeong** [reply]
- **한 클래스**에서 빌터 패턴과 정적 팩토리 메서드를 함께 사용한 부분에 대한 내용인가요?
   - 대답) 네! 맞습니다~ 현재 제가 맡은 도메인에서는 entity 클래스에 정적 팩토리 메서드와 빌더패턴을 같이 쓰고있어서 위와 같은 고민을 하게 되었습니다
- dto쪽에 빌더 패턴을 사용했을 때 이슈 발생할 여지가 있음
   - 대답) 오?! 어떤 문제가 발생할 수 있는지 설명해 주실 수 있나요?? 궁금합니다!! 저도 참고할래요~~

**Reply by locodingve** [reply]
지금 notification 도메인에 예시가 있습니다. 
지금 저희 공지 내용 종류는 loudspeaker와 follow 알림 정보가 있습니다.

우선 notification request dto는 아래와 같습니다. (언어: TypeScript 버전)
``` 
{
  ...
  type: 'LOUDSPEAKER | FOLLOW'
  follow?: {
    ...
  },
  loudSpeaker?: {
     ...
  },
  ...
}
```

제가 말씀드리는 싶은 부분은 loudspeaker내용이면 loudspeaker 객체의 내용을 담고, follow 정보이면 follow 객체에 내용을 담습니다. 
타입에 따라 정적메소드를 이용하며, 정확한 객체를 생성하고 싶어서 입니다. 
왜냐하면 notification 도메인 쪽에서는 하나의 notification이라는 객체에 'loudspeaaker와 follow 정보'를 둘 다 담을 수 있기 때문에, 
정보에 따라(=loudspeaker or follow) dto에 담는 과정과 entity에 담는 과정이 굉장히 중요합니다. 
그래서 dto와 entity에서 두 번 정확하게 유효성 검사를 하고 객체를 생성해야 한다고 생각합니다.
